pkg/deploy/controller.go

```diff
@@ -303,10 +303,9 @@ func (w *watcher) delete(path string) error {
 		return err
 	}
 
-	// ensure that the addon is completely removed before deleting the objectSet,
-	// so return when err == nil, otherwise pods may get stuck terminating
+	// ensure that the addon is completely removed before deleting the objectSet
 	w.recorder.Eventf(&addon, corev1.EventTypeNormal, "DeletingManifest", "Deleting manifest at %q", path)
-	if err := w.addons.Delete(addon.Namespace, addon.Name, &metav1.DeleteOptions{}); err == nil || !apierrors.IsNotFound(err) {
+	if err := w.addons.Delete(addon.Namespace, addon.Name, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
```

Comment:
I am a little concerned about this, but only because the Wrangler OnRemove handler can behave in unexpected ways.

The fact that we were returning early here if the delete succeeds OR returns an error other than not-found makes me suspect that we are relying on the controller enqueueing and re-executing this handler multiple times. Otherwise the apply and file delete down below would never happen.

A lot has changed since this code and the comment associated with it were added. For example, we did not actually track GVKs at the time. I would almost suggest that in addition to this change, we move this down below to not remove the AddOn until AFTER the apply and file delete have succeeded.

Reply:
Sounds good, I changed the order as suggested.

Reply:
Thanks! This makes more sense to me. I think that given we are now tracking GVKs on the AddOn, we should not delete it until after the apply succeeds - otherwise we risk losing track of some resources, as you noted.